File: process_excel.py
import pandas as pd
import numpy as np
import re
import os
from openpyxl import Workbook
from openpyxl.styles import Font, PatternFill
from datetime import datetime

def parse_price(val):
    if val is None or pd.isna(val) or val == '':
        return None
    if isinstance(val, (int, float)):
        return float(val)
    str_val = str(val).strip()
    if not str_val:
        return None
    # cleanStr = strVal.replace(/[^\d,-]/g, '').replace(',', '.');
    clean_str = re.sub(r'[^\d,-]', '', str_val).replace(',', '.')
    try:
        return float(clean_str)
    except ValueError:
        return None

def process_and_diagnose_sheet(input_path, mappings, class_map, selected_sheet, filter_config):
    # Read Excel using pandas
    df = pd.read_excel(input_path, sheet_name=selected_sheet, dtype=str)
    
    # We want to match headers
    col_code = mappings.get('kode')
    col_name = mappings.get('nama')
    col_class = mappings.get('kelas')
    col_price = mappings.get('harga')
    
    if not all(col in df.columns for col in [col_code, col_name, col_class, col_price]):
        raise Exception("Pemetaan kolom tidak valid.")
        
    # Check if we have a filter
    use_filter = filter_config and filter_config.get('column') and len(filter_config.get('values', [])) > 0
    col_filter = filter_config.get('column') if use_filter else None
    filter_values = [str(v).upper() for v in filter_config.get('values', [])] if use_filter else []
    
    rejected_rows_sample = []
    REJECTED_SAMPLE_SIZE = 10
    total_rows_read = 0
    total_rows_filtered = 0
    
    grouped = {}
    
    for idx, row in df.iterrows():
        total_rows_read += 1
        
        # Apply filter
        if use_filter and col_filter in df.columns:
            cell_value = str(row[col_filter]).strip() if not pd.isna(row[col_filter]) else ''
            if cell_value.upper() not in filter_values:
                if len(rejected_rows_sample) < REJECTED_SAMPLE_SIZE:
                    rejected_rows_sample.append({
                        'kode': str(row[col_code]).strip() if not pd.isna(row[col_code]) else '',
                        'nama': str(row[col_name]).strip() if not pd.isna(row[col_name]) else '',
                        'alasan': f"Nilai '{cell_value}' tidak cocok filter"
                    })
                total_rows_filtered += 1
                continue
                
        code = str(row[col_code]).strip() if not pd.isna(row[col_code]) else ''
        name = str(row[col_name]).strip() if not pd.isna(row[col_name]) else ''
        
        if not code or not name:
            continue
            
        kelas_from_file = str(row[col_class]).strip().upper() if not pd.isna(row[col_class]) else ''
        price_raw = row[col_price]
        price = parse_price(price_raw)
        
        key = f"{code}|{name}"
        if key not in grouped:
            grouped[key] = {
                'code': code,
                'name': name,
                'OPD': None,
                'ED': None,
                'KELAS 3': None,
                'KELAS 2': None,
                'KELAS 1': None,
                'VIP': None,
                'VVIP': None
            }
            
        target_col = class_map.get(kelas_from_file)
        if target_col and target_col != 'ignore':
            grouped[key][target_col] = price
            
    all_accepted_rows = list(grouped.values())
    accepted_rows_sample = all_accepted_rows[:10]
    
    summary = {
        'totalRowsRead': total_rows_read,
        'totalRowsFiltered': total_rows_filtered,
        'totalRowsProcessed': total_rows_read - total_rows_filtered,
        'uniqueItemCount': len(all_accepted_rows)
    }
    
    return {
        'summary': summary,
        'rejectedRowsSample': rejected_rows_sample,
        'acceptedRowsSample': accepted_rows_sample,
        'allAcceptedRows': all_accepted_rows
    }

def create_final_excel(rows, output_dir):
    wb = Workbook()
    ws = wb.active
    ws.title = "Buku Tarif LAB"
    
    headers = [
        'Kode', 'Nama Pemeriksaan', 'OPD', 'ED', 
        'KELAS 3', 'KELAS 2', 'KELAS 1', 'VIP', 'VVIP'
    ]
    
    ws.append(headers)
    
    # Column styling
    header_font = Font(bold=True)
    header_fill = PatternFill(start_color="D3D3D3", end_color="D3D3D3", fill_type="solid")
    
    for col_idx in range(1, len(headers) + 1):
        cell = ws.cell(row=1, column=col_idx)
        cell.font = header_font
        cell.fill = header_fill
        
    # Append rows
    for r in rows:
        row_data = [
            r.get('code', ''),
            r.get('name', ''),
            r.get('OPD'),
            r.get('ED'),
            r.get('KELAS 3'),
            r.get('KELAS 2'),
            r.get('KELAS 1'),
            r.get('VIP'),
            r.get('VVIP')
        ]
        ws.append(row_data)
        
    # Format number columns (columns 3 to 9) to standard number format #,##0
    for row_idx in range(2, len(rows) + 2):
        for col_idx in range(3, 10):
            cell = ws.cell(row=row_idx, column=col_idx)
            if cell.value is not None:
                cell.number_format = '#,##0'
                
    # Auto-adjust column widths
    column_widths = {
        1: 15, 2: 45, 3: 15, 4: 15, 5: 15, 6: 15, 7: 15, 8: 15, 9: 15
    }
    for col_idx, width in column_widths.items():
        ws.column_dimensions[ws.cell(row=1, column=col_idx).column_letter].width = width
        
    timestamp = datetime.now().strftime('%Y-%m-%dT%H-%M-%S')
    out_file_name = f"Buku_Tarif_LAB_{timestamp}.xlsx"
    out_path = os.path.join(output_dir, out_file_name)
    wb.save(out_path)
    return out_path

File: processExcel.js
const Excel = require('exceljs');
const path = require('path');

function parsePrice(val) {
  if (val === null || val === undefined || val === '') return null;
  if (typeof val === 'number') return val;
  const strVal = String(val).trim();
  if (!strVal) return null;
  const cleanStr = strVal.replace(/[^\d,-]/g, '').replace(',', '.');
  const parsed = parseFloat(cleanStr);
  return isNaN(parsed) ? null : parsed;
}

async function processAndDiagnoseSheet(inputPath, mappings, classMap, selectedSheet, filterConfig) {
  const workbook = new Excel.Workbook();
  await workbook.xlsx.readFile(inputPath);
  
  const sheet = workbook.getWorksheet(selectedSheet);
  if (!sheet) throw new Error(`Sheet "${selectedSheet}" tidak ditemukan.`);

  const headerRow = sheet.getRow(1);
  const headers = {};
  headerRow.eachCell((cell, col) => { if (cell.text) headers[cell.text.trim()] = col; });

  const colCode = headers[mappings.kode];
  const colName = headers[mappings.nama];
  const colClass = headers[mappings.kelas];
  const colPrice = headers[mappings.harga];
  if (!colCode || !colName || !colClass || !colPrice) throw new Error("Pemetaan kolom tidak valid.");
  
  const useFilter = filterConfig && filterConfig.column && filterConfig.values.length > 0;
  const colFilter = useFilter ? headers[filterConfig.column] : -1;
  const includedValues = useFilter ? filterConfig.values.map(v => v.toUpperCase()) : [];

  const grouped = {};
  const rejectedRowsSample = [];
  const REJECTED_SAMPLE_SIZE = 10;
  let totalRowsRead = 0;
  let totalRowsFiltered = 0;

  sheet.eachRow((row, rowNum) => {
    if (rowNum === 1) return;
    totalRowsRead++;

    if (useFilter && colFilter !== -1) {
      const cellValue = (row.getCell(colFilter).text || '').trim();
      if (!includedValues.includes(cellValue.toUpperCase())) {
        if (rejectedRowsSample.length < REJECTED_SAMPLE_SIZE) {
            rejectedRowsSample.push({ 
                kode: (row.getCell(colCode).text || '').trim(), 
                nama: (row.getCell(colName).text || '').trim(),
                alasan: `Nilai '${cellValue}' tidak cocok filter` 
            });
        }
        totalRowsFiltered++;
        return;
      }
    }

    const code = (row.getCell(colCode).text || '').trim();
    const name = (row.getCell(colName).text || '').trim();
    if (!code || !name) return;

    const kelasFromFile = (row.getCell(colClass).text || '').trim().toUpperCase();
    const price = parsePrice(row.getCell(colPrice).value);
    const key = code + '|' + name;
    if (!grouped[key]) grouped[key] = { code, name };
    const targetCol = classMap[kelasFromFile];
    if (targetCol && targetCol !== 'ignore') grouped[key][targetCol] = price;
  });

  const allAcceptedRows = Object.values(grouped);
  const acceptedRowsSample = allAcceptedRows.slice(0, 10);

  const summary = {
      totalRowsRead,
      totalRowsFiltered,
      totalRowsProcessed: totalRowsRead - totalRowsFiltered,
      uniqueItemCount: allAcceptedRows.length,
  };

  return { summary, rejectedRowsSample, acceptedRowsSample, allAcceptedRows };
}

async function createFinalExcel(rows) {
    const outWb = new Excel.Workbook();
    const outSheet = outWb.addWorksheet('Buku Tarif LAB');
    outSheet.columns = [
        { header: 'Kode', key: 'code', width: 15 }, { header: 'Nama Pemeriksaan', key: 'name', width: 45 },
        { header: 'OPD', key: 'OPD', width: 15, style: { numFmt: '#,##0' } }, { header: 'ED', key: 'ED', width: 15, style: { numFmt: '#,##0' } },
        { header: 'KELAS 3', key: 'KELAS 3', width: 15, style: { numFmt: '#,##0' } }, { header: 'KELAS 2', key: 'KELAS 2', width: 15, style: { numFmt: '#,##0' } },
        { header: 'KELAS 1', key: 'KELAS 1', width: 15, style: { numFmt: '#,##0' } }, { header: 'VIP', key: 'VIP', width: 15, style: { numFmt: '#,##0' } },
        { header: 'VVIP', key: 'VVIP', width: 15, style: { numFmt: '#,##0' } },
    ];
    outSheet.getRow(1).font = { bold: true };
    outSheet.getRow(1).fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FFD3D3D3' } };
    outSheet.addRows(rows);

    const outputDir = path.join(__dirname, 'output');
    const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
    const outFileName = `Buku_Tarif_LAB_${timestamp}.xlsx`;
    const outPath = path.join(outputDir, outFileName);
    await outWb.xlsx.writeFile(outPath);
    return outPath;
}

module.exports = { processAndDiagnoseSheet, createFinalExcel };